releases: add v4.0.9 release notes (#4294)
pingcap/docs: releases/release-4.0.9.md

    - Fix auto-id allocation failed because of transaction's write-conflict retry. [#21079](https://github.com/pingcap/tidb/pull/21079)
    - Fix The JSON Data can not import to TiDB correctly by `load data` [#21074](https://github.com/pingcap/tidb/pull/21074)
    - Fix up displaying executor runtime info & Add runtime information for DML in explain analyze [#21066](https://github.com/pingcap/tidb/pull/21066)
    - Deprecate `enable-streaming`. [#21055](https://github.com/pingcap/tidb/pull/21055)
    - Avoid unnecessary warnings/errors when folding constants in shortcut-able expressions [#21040](https://github.com/pingcap/tidb/pull/21040)
    - TiDB now provides an error for statements that use the syntax `LOCK IN SHARE MODE`. This helps alert users that this feature is not implemented by TiDB, but a workaround to restore the previous behavior is available by setting `tidb_enable_noop_functions=1`. [#21005](https://github.com/pingcap/tidb/pull/21005)
    - Set correct default value for new added enum column. [#20998](https://github.com/pingcap/tidb/pull/20998)
    - Allow set GOGC in config file. [#20922](https://github.com/pingcap/tidb/pull/20922)

+ TiKV

>     - Fix a bug that Coprocessor may return wrong result when there are > 255 columns. [#9131](https://github.com/tikv/tikv/pull/9131)

Reviewer: All the items that starts with `Fix` should probably be moved to bug fix.

Reviewer 2: moved

    - Fix an issue that region merge may cause data loss during network partition. [#9108](https://github.com/tikv/tikv/pull/9108)
    - Enable dynamically change config pessimistic-txn.pipelined. [#9100](https://github.com/tikv/tikv/pull/9100)
    - Running checksum in BR and Lightning should have less influence on the cluster performance. [#9098](https://github.com/tikv/tikv/pull/9098)

>     - Revert "encryption: Turn file dictionary into log style" [#9097](https://github.com/tikv/tikv/pull/9097)

Reviewer: This and "Turn file dictionary into log style" can be removed.

Reviewer 2: done

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

Release date: December 3, 2020

TiDB version: 4.0.9

## New Features

+ TiFlash

>     - Support multi-disks for storing latest data [#1159](https://github.com/pingcap/tics/pull/1159)

Reviewer 3: ```suggestion
    - Support storing the latest data of the storage engine on multiple disks. [#1159](https://github.com/pingcap/tics/pull/1159)
```

## Improvements

+ TiDB

    - Ban (index) merge join heuristically when convert equal conditions to other conditions. [#21146](https://github.com/pingcap/tidb/pull/21146)
    - Differentiate types for user variables [#21107](https://github.com/pingcap/tidb/pull/21107)
    - Allow set GOGC in config file. [#20922](https://github.com/pingcap/tidb/pull/20922)

+ TiKV

    - Add tag to split command to trace its source [#8936](https://github.com/tikv/tikv/pull/8936)

+ TiFlash

>     - Implement `batch_read_index` to reduce latency about replica read. [#1236](https://github.com/pingcap/tics/pull/1236)

Reviewer 3: ```suggestion
    - Reduce latency about replica read. [#1236](https://github.com/pingcap/tics/pull/1236)
```

    - Refine TiFlash's error message [#1231](https://github.com/pingcap/tics/pull/1231)
    - Limit the memory usage of cache data under huge volume data. [#1229](https://github.com/pingcap/tics/pull/1229)
    - [#1227](https://github.com/pingcap/tics/pull/1227)

>     - Add metric "number of handling cop task". [#1183](https://github.com/pingcap/tics/pull/1183)

Reviewer 3: ```suggestion
    - Add metric about the number of handling coprocessor tasks. [#1183](https://github.com/pingcap/tics/pull/1183)
```

    - Fix a bug may cause analyze statement panic when using the latin1 character set. [#9082](https://github.com/tikv/tikv/pull/9082)
    - Fix cast decimal as time expr get wrong result when deal with numbers argument [#9031](https://github.com/tikv/tikv/pull/9031)
    - Fix the bug that lightning fails to ingest sst files to tikv with importer/local backend when TDE is enabled [#8995](https://github.com/tikv/tikv/pull/8995)
    - Config: fix invalid advertise-status-addr [#9036](https://github.com/tikv/tikv/pull/9036)
    - Fix the issue that reports a key-exist error when a key is locked and deleted by a committed transaction. [#8930](https://github.com/tikv/tikv/pull/8930)

+ PD

    - Fix the issue that the leader role does not take effect when using the replacement rule in some cases [#3208](https://github.com/pingcap/pd/pull/3208)
    - Fix the bug that `trace-region-flow` will be accidentally changed to `false`. [#3120](https://github.com/pingcap/pd/pull/3120)

+ TiFlash

    - Fix the problem that `INFORMATION_SCHEMA.CLUSTER_HARDWARE` may contain information about disks not in use. [#1248](https://github.com/pingcap/tics/pull/1248)

>     - Fix the issue that the memory consumption statistic of Delta Cache is smaller than actual usage. [#1212](https://github.com/pingcap/tics/pull/1212)

Reviewer 3: ```suggestion
    - Fix the issue that the memory consumption statistic of Delta Cache is smaller than actual the usage. [#1212](https://github.com/pingcap/tics/pull/1212)
```

    - Change the default value of `apply-max-batch-size` and `store-max-batch-size` to 1024 [#9020](https://github.com/tikv/tikv/pull/9020)
    - Add config for max-background-flushes [#8947](https://github.com/tikv/tikv/pull/8947)
    - Enable the unified read pool for the storage module by default. [#8887](https://github.com/tikv/tikv/pull/8887)

+ PD

    - Dashboard: update to v2020.11.26.1 [#3219](https://github.com/pingcap/pd/pull/3219)
    - Check cluster version on stores become tombstone [#3213](https://github.com/pingcap/pd/pull/3213)
    - The store of lower version is no longer allowed to change from `Tombstone` back to `Up`. [#3206](https://github.com/pingcap/pd/pull/3206)
    - Fix a bug that service safe points with  infinite TTL may disappear. [#3143](https://github.com/pingcap/pd/pull/3143)

+ Tools

    - TiCDC

        * Metrics: add alert for tikv hibernate regions [#1120](https://github.com/pingcap/ticdc/pull/1120)
        * Fix the bug that multiple owners could exist when owner campaign key is deleted [#1104](https://github.com/pingcap/ticdc/pull/1104)

>         * Kafka_consumer: fix wrong timezone when 'TZ' env is not set and downstream is mysql [#1095](https://github.com/pingcap/ticdc/pull/1095)

Reviewer 4: ```suggestion
        * Kafka_consumer: fix wrong timezone when 'TZ' env is not set and downstream is mysql [#1095](https://github.com/pingcap/ticdc/pull/1095)
        * Reduce memory usage in schema storage by avoiding table info replication [#1127](https://github.com/pingcap/ticdc/pull/1127)

# TiDB-binlog

## Improvements

* Exit drainer process with non-zero code if upstream PD is down or apply DDL/DML to downstream failed [#1012](https://github.com/pingcap/tidb-binlog/pull/1012)

# BR

## Improvements

* BR no longer accepts --checksum false in command line, which did not disable checksum. The correct usage always include the = sign: --checksum=false [#588](https://github.com/pingcap/br/pull/588)
* Support change PD config temporary [#596](https://github.com/pingcap/br/pull/596)
* Support analyze table after restore [#622](https://github.com/pingcap/br/pull/622)
* Retry for read index not ready and proposal in merging mode [#626](https://github.com/pingcap/br/pull/626)

## Bug Fixes

* Fix special characters in S3 secret access keys [#617](https://github.com/pingcap/br/pull/617)

# Dumpling

## Improvements

* Retry dumping on failed chunks [#182](https://github.com/pingcap/dumpling/pull/182)
* Support configuring both -F and -r arguments at the same time [#177](https://github.com/pingcap/dumpling/pull/177)
* Exclude system databases in --filter parameter by default [#194](https://github.com/pingcap/dumpling/pull/194)
* Support --transactional-consistency parameter and support rebuild mysql connections during retry [#199](https://github.com/pingcap/dumpling/pull/199)

## Bug Fixes

* fix the problem that dumpling may get blocked when its connection to database server is closed [#190](https://github.com/pingcap/dumpling/pull/190)

# TiDB-Lightning

## Improvements

* Filter out all system schemas by default [#459](https://github.com/pingcap/tidb-lightning/pull/459)
* Support set a default value for auto random primary key for local/importer backend [#457](https://github.com/pingcap/tidb-lightning/pull/457)
* Use range properties to make range split more precise with local backend [#422](https://github.com/pingcap/tidb-lightning/pull/422)
* `tikv-importer.region-split-size`, `mydumper.read-block-size`, `mydumper.batch-size` and `mydumper.max-region-size` can now accept human-readable format in the form "2.5 GiB" [#471](https://github.com/pingcap/tidb-lightning/pull/471)

## Bug Fixes

* Fix a bug about encoding data with wrong field info [#437](https://github.com/pingcap/tidb-lightning/pull/437)
* Fix a bug that gc life time ttl does not take effect [#448](https://github.com/pingcap/tidb-lightning/pull/448)
* Fix a bug that causes panic when manually stops Lightning import in localbackend mode [#484](https://github.com/pingcap/tidb-lightning/pull/484)

```

Reviewer 2: Addd in https://github.com/pingcap/docs/pull/4294/commits/68de6db8c7e09deeaae3f6a7db45b60ed16f0ef7.

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

> Release date: December 3, 2020

Reviewer 2: ```suggestion
Release date: December 10, 2020
```

    - Fix the bug that lightning fails to ingest sst files to tikv with importer/local backend when TDE is enabled [#8995](https://github.com/tikv/tikv/pull/8995)
    - Config: fix invalid advertise-status-addr [#9036](https://github.com/tikv/tikv/pull/9036)
    - Fix the issue that reports a key-exist error when a key is locked and deleted by a committed transaction. [#8930](https://github.com/tikv/tikv/pull/8930)
    - Add check to avoid corruption due to RocksDB block cache error #8243. Also disable `force-consistency-checks` by default to improve performance. [#9029](https://github.com/tikv/tikv/pull/9029)

+ PD

    - Fix the issue that the leader role does not take effect when using the replacement rule in some cases [#3208](https://github.com/pingcap/pd/pull/3208)
    - Fix the bug that `trace-region-flow` will be accidentally changed to `false`. [#3120](https://github.com/pingcap/pd/pull/3120)
    - Fix a bug that service safe points with infinite TTL may disappear. [#3143](https://github.com/pingcap/pd/pull/3143)

+ TiFlash

    - Fix the problem that `INFORMATION_SCHEMA.CLUSTER_HARDWARE` may contain information about disks not in use

>     - Fix the issue that the memory consumption statistic of Delta Cache is smaller than actual the usage

Reviewer 3: ```suggestion
    - Fix the issue that the memory consumption statistic of Delta Cache is smaller than actual the usage
    - Fix memory leak about thread info statistics
```

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

Release date: December 10, 2020

TiDB version: 4.0.9

## Compatibility changes

+ TiDB

    - Deprecate `enable-streaming` [#21055](https://github.com/pingcap/tidb/pull/21055)

## New Features

+ TiFlash

>     - Support storing the latest data of the storage engine on multiple disks

Reviewer 3: Finally, this is an experimental feature in the release-4.0 branch. Should we add some tips in the release note? Like adding `(experimental)` at the end of the line. @TomShawn

Reviewer 2: Of course.
```suggestion
    - Support storing the latest data of the storage engine on multiple disks (experimental)
```

    - TiDB now provides an error for statements that use the syntax `LOCK IN SHARE MODE`. This helps alert users that this feature is not implemented by TiDB, but a workaround to restore the previous behavior is available by setting `tidb_enable_noop_functions=1`. [#21005](https://github.com/pingcap/tidb/pull/21005)
    - Avoid unnecessary warnings/errors when folding constants in shortcut-able expressions [#21040](https://github.com/pingcap/tidb/pull/21040)
    - Raise an error when preparing the `load data` statement. [#21199](https://github.com/pingcap/tidb/pull/21199)
    - Ignore integer zero-fill size attribute when changing the column types [#20986](https://github.com/pingcap/tidb/pull/20986)
    - Add runtime information for DML in explain analyze [#21066](https://github.com/pingcap/tidb/pull/21066)
    - Disallow multi-updates on the primary key (i.e. updating primary key of the same table with different alias) [#21113](https://github.com/pingcap/tidb/pull/21113)
    - Add metric for connection idle time [#21301](https://github.com/pingcap/tidb/pull/21301)

+ TiKV

    - Add tag to split command to trace its source [#8936](https://github.com/tikv/tikv/pull/8936)

> 

Reviewer 5: ```suggestion
```suggestion
+ PD
    - Check cluster version on stores become tombstone [#3213](https://github.com/pingcap/pd/pull/3213)
    - The store of lower version is no longer allowed to change from `Tombstone` back to `Up`. [#3206](https://github.com/pingcap/pd/pull/3206)
```
```

+ PD

    - Check cluster version on stores become tombstone [#3213](https://github.com/pingcap/pd/pull/3213)
    - The store of lower version is no longer allowed to change from `Tombstone` back to `Up`.[#3206](https://github.com/pingcap/pd/pull/3206)
    - Dashboard: update to v2020.11.26.1 [#3219](https://github.com/pingcap/pd/pull/3219)

+ TiFlash

    - Reduce the latency of replica reads [#1236](https://github.com/pingcap/tics/pull/1236)
    - Refine TiFlash's error message [#1231](https://github.com/pingcap/tics/pull/1231)
    - Limit the memory usage of cache data under huge volume data. [#1229](https://github.com/pingcap/tics/pull/1229)

>     - [#1227](https://github.com/pingcap/tics/pull/1227)

Reviewer 3: ```suggestion
```

+ TiKV

    - Fix a bug that Coprocessor may return wrong result when there are > 255 columns. [#9131](https://github.com/tikv/tikv/pull/9131)
    - Fix an issue that region merge may cause data loss during network partition. [#9108](https://github.com/tikv/tikv/pull/9108)
    - Fix a bug may cause analyze statement panic when using the latin1 character set. [#9082](https://github.com/tikv/tikv/pull/9082)
    - Fix cast decimal as time expr get wrong result when deal with numbers argument [#9031](https://github.com/tikv/tikv/pull/9031)
    - Fix the bug that lightning fails to ingest sst files to tikv with importer/local backend when TDE is enabled [#8995](https://github.com/tikv/tikv/pull/8995)
    - Config: fix invalid advertise-status-addr [#9036](https://github.com/tikv/tikv/pull/9036)
    - Fix the issue that reports a key-exist error when a key is locked and deleted by a committed transaction. [#8930](https://github.com/tikv/tikv/pull/8930)

+ PD

    - Fix the issue that the leader role does not take effect when using the replacement rule in some cases [#3208](https://github.com/pingcap/pd/pull/3208)

>     - Fix the bug that `trace-region-flow` will be accidentally changed to `false`. [#3120](https://github.com/pingcap/pd/pull/3120)

Reviewer 5: ```suggestion
    - Fix the bug that `trace-region-flow` will be accidentally changed to `false`. [#3120](https://github.com/pingcap/pd/pull/3120)
```
```suggestion
    - Fix the bug that `trace-region-flow` will be accidentally changed to `false`. [#3120](https://github.com/pingcap/pd/pull/3120)
    - Fix a bug that service safe points with infinite TTL may disappear. [#3143](https://github.com/pingcap/pd/pull/3143)
```

+ Tools

    + Backup & Restore (BR)

        - Fix special characters in S3 secret access keys [#617](https://github.com/pingcap/br/pull/617)

    + TiCDC

        - Fix the bug that multiple owners could exist when owner campaign key is deleted [#1104](https://github.com/pingcap/ticdc/pull/1104)
        - Kafka_consumer: fix wrong timezone when 'TZ' env is not set and downstream is mysql [#1095](https://github.com/pingcap/ticdc/pull/1095)

    + Dumpling

        - Fix the problem that dumpling may get blocked when its connection to database server is closed [#190](https://github.com/pingcap/dumpling/pull/190)

    + TiDB Lightning

        - Fix a bug about encoding data with wrong field info [#437](https://github.com/pingcap/tidb-lightning/pull/437)
        - Fix a bug that gc life time ttl does not take effect [#448](https://github.com/pingcap/tidb-lightning/pull/448)
        - Fix a bug that causes panic when manually stops Lightning import in Local-backend mode [#484](https://github.com/pingcap/tidb-lightning/pull/484)

## Others

+ TiKV

>     - Enable dynamically change config pessimistic-txn.pipelined. [#9100](https://github.com/tikv/tikv/pull/9100)

Reviewer: ```suggestion
    - Support changing config pessimistic-txn.pipelined dynamically. [#9100](https://github.com/tikv/tikv/pull/9100)
```

>         - Kafka_consumer: fix wrong timezone when 'TZ' env is not set and downstream is mysql [#1095](https://github.com/pingcap/ticdc/pull/1095)

Reviewer 4: ```suggestion
```

    - Running checksum in BR and Lightning should have less influence on the cluster performance. [#9098](https://github.com/tikv/tikv/pull/9098)

>     - Add a metrics for ingest sst errors [#9096](https://github.com/tikv/tikv/pull/9096)

Reviewer: ```suggestion
    - Add a metrics for ingesting sst errors [#9096](https://github.com/tikv/tikv/pull/9096)
```

    - Prevent hibernation when there are still peers catching up logs [#9093](https://github.com/tikv/tikv/pull/9093)
    - Increase the success rate of pipelined pessimistic locking. [#9086](https://github.com/tikv/tikv/pull/9086)

>     - Add check to avoid corruption due to RocksDB block cache error #8243. Also disable `force-consistency-checks` by default to improve performance. [#9029](https://github.com/tikv/tikv/pull/9029)

Reviewer: This is a bug fix, others are improvment.

>     - Check cluster version on stores become tombstone [#3213](https://github.com/pingcap/pd/pull/3213)
>     - The store of lower version is no longer allowed to change from `Tombstone` back to `Up`. [#3206](https://github.com/pingcap/pd/pull/3206)
>     - Fix a bug that service safe points with  infinite TTL may disappear. [#3143](https://github.com/pingcap/pd/pull/3143)

Reviewer 5: ```suggestion
    - Check cluster version on stores become tombstone [#3213](https://github.com/pingcap/pd/pull/3213)
    - The store of lower version is no longer allowed to change from `Tombstone` back to `Up`. [#3206](https://github.com/pingcap/pd/pull/3206)
    - Fix a bug that service safe points with  infinite TTL may disappear. [#3143](https://github.com/pingcap/pd/pull/3143)
```
```suggestion
```

    - Fix the issue that Region Merge might cause data loss during network partition [#9108](https://github.com/tikv/tikv/pull/9108)
    - Fix the issue that the `ANALYZE` statement might cause panic when using the `latin1` character set [#9082](https://github.com/tikv/tikv/pull/9082)
    - Fix the wrong results returned when converting the numeric type to the time type [#9031](https://github.com/tikv/tikv/pull/9031)
    - Fix the bug that TiDB Lightning fails to ingest SST files to TiKV with the Importer-backend or Local-backend when
 Transparent Data Encryption (TDE) is enabled [#8995](https://github.com/tikv/tikv/pull/8995)
    - Fix the invalid `advertise-status-addr` value `0.0.0.0` [#9036](https://github.com/tikv/tikv/pull/9036)
    - Fix the issue that an error is returned indicating that a key exists when this key is locked and deleted in a committed transaction [#8930](https://github.com/tikv/tikv/pull/8930)
    - Fix the issue that the RocksDB mapping error causes data corruption [#9029](https://github.com/tikv/tikv/pull/9029)

> 

Reviewer 6: ```suggestion
- Fix the bug that follower read may get stale data after transferring leader [#9240](https://github.com/tikv/tikv/pull/9240)

```

    - Update Dashboard to `v2020.11.26.1` [#3219](https://github.com/pingcap/pd/pull/3219)

+ TiFlash

    - Reduce the latency of replica reads
    - Refine TiFlash's error messages
    - Limit the memory usage of cache data when the data volume is huge
    - Add a monitoring metric for the number of coprocessor tasks being handled

+ Tools

    + Backup & Restore (BR)

        - Disallow the ambiguous `--checksum false` argument in the command line, which does not disable checksum. Only `--checksum=false` is accepted. [#588](https://github.com/pingcap/br/pull/588)
        - Support changing the PD configuration temporarily [#596](https://github.com/pingcap/br/pull/596)
        - Support analyzing tables after restore [#622](https://github.com/pingcap/br/pull/622)
        - Retry for the `read index not ready` and `proposal in merging mode` errors [#626](https://github.com/pingcap/br/pull/626)

    + TiCDC

        - Add an alert for enabling TiKV's Hibernate Region feature [#1120](https://github.com/pingcap/ticdc/pull/1120)
        - Reduce memory usage in the schema storage [#1127](https://github.com/pingcap/ticdc/pull/1127)

> 

Reviewer 7: ```suggestion
        - Add unified sorter, which will accelerate replication when incremental scan data size is large (experimental) [#1122](https://github.com/pingcap/ticdc/pull/1122)
        - Support to configure max message size and max message batch in Open Protocol message (Kafka sink only) [#1079](https://github.com/pingcap/ticdc/pull/1079)
        
```

    - Fix the issue that Region Merge might cause data loss during network partition [#9108](https://github.com/tikv/tikv/pull/9108)
    - Fix the issue that the `ANALYZE` statement might cause panic when using the `latin1` character set [#9082](https://github.com/tikv/tikv/pull/9082)
    - Fix the wrong results returned when converting the numeric type to the time type [#9031](https://github.com/tikv/tikv/pull/9031)
    - Fix the bug that TiDB Lightning fails to ingest SST files to TiKV with the Importer-backend or Local-backend when
 Transparent Data Encryption (TDE) is enabled [#8995](https://github.com/tikv/tikv/pull/8995)
    - Fix the invalid `advertise-status-addr` value `0.0.0.0` [#9036](https://github.com/tikv/tikv/pull/9036)
    - Fix the issue that an error is returned indicating that a key exists when this key is locked and deleted in a committed transaction [#8930](https://github.com/tikv/tikv/pull/8930)
    - Fix the issue that the RocksDB mapping error causes data corruption [#9029](https://github.com/tikv/tikv/pull/9029)

> - Fix the bug that Follower Read might return stale data after the leader is transferred [#9240](https://github.com/tikv/tikv/pull/9240)

Reviewer 2: ```suggestion
    - Fix the bug that Follower Read might return stale data after the leader is transferred [#9240](https://github.com/tikv/tikv/pull/9240)
```

+ PD

    - Fix the issue that the leader roles specified using placement rules do not take effect in some cases [#3208](https://github.com/pingcap/pd/pull/3208)
    - Fix the issue that the `trace-region-flow` value is unexpectedly set to `false` [#3120](https://github.com/pingcap/pd/pull/3120)
    - Fix a bug that the service safepoint with infinite Time To Live (TTL) does not work [#3143](https://github.com/pingcap/pd/pull/3143)

+ TiFlash

    - Fix the issue that `INFORMATION_SCHEMA.CLUSTER_HARDWARE` might contain the information of disks that are not in use
    - Fix the issue that the estimate on memory usage of Delta Cache is smaller than the actual usage
    - Fix the memory leak caused by thread information statistics

+ Tools

    + Backup & Restore (BR)

        - Fix the failure caused by special characters in S3 secret access keys [#617](https://github.com/pingcap/br/pull/617)

    + TiCDC

>         - Fix the issue that multiple owners might exist when the owner campaign key is deleted [#1104](https://github.com/pingcap/ticdc/pull/1104)

Reviewer 7: ```suggestion
        - Fix the issue that multiple owners might exist when the owner campaign key is deleted [#1104](https://github.com/pingcap/ticdc/pull/1104)
        - Fix a bug that TiCDC could fail to continue replicating when a TiKV crashes or recovers from a crash, the bug exists in v4.0.8 only [#1198](https://github.com/pingcap/ticdc/pull/1198)
         - Fix task status flushed to etcd too many before a table is initialized [#1191](https://github.com/pingcap/ticdc/pull/1191)
         - Fix schema storage may meet error because of update latency or GC too fast [#1114](https://github.com/pingcap/ticdc/pull/1114)
         - Fix schema storage costs too much memory when DDL is frequent [#1127](https://github.com/pingcap/ticdc/pull/1127)
         - Fix goroutine leak when changefeed is paused or stopped [#1075](https://github.com/pingcap/ticdc/pull/1075)
         - Increase the max retry timeout to 60s Kafka produce to prevent replication interruption because of service or network jitter in downstream Kafaka [#1118](https://github.com/pingcap/ticdc/pull/1118)
         - Fix the bug that Kafka batch size does not work [#1112](https://github.com/pingcap/ticdc/pull/1112)
```

Reviewer 7: PTAL @TomShawn Many thanks

Reviewer 2: No problem. Could you also provide the Chinese version in https://github.com/pingcap/docs-cn/pull/4976?

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

Release date: December 3, 2020

TiDB version: 4.0.9

## New Features

+ TiFlash

    - Support storing the latest data of the storage engine on multiple disks [#1159](https://github.com/pingcap/tics/pull/1159)

## Improvements

+ TiDB

>     - Ban (index) merge join heuristically when convert equal conditions to other conditions. [#21146](https://github.com/pingcap/tidb/pull/21146)
>     - Differentiate types for user variables [#21107](https://github.com/pingcap/tidb/pull/21107)
>     - Allow set GOGC in config file. [#20922](https://github.com/pingcap/tidb/pull/20922)

Reviewer 8: ```suggestion
    - Ban (index) merge join heuristically when converting equal conditions to other conditions. [#21146](https://github.com/pingcap/tidb/pull/21146)
    - Differentiate types for user variables [#21107](https://github.com/pingcap/tidb/pull/21107)
    - Allow set GOGC in the config file. [#20922](https://github.com/pingcap/tidb/pull/20922)
    - Make dump binary time more compatible with MySQL [#21135](https://github.com/pingcap/tidb/pull/21135)
    - TiDB now provides an error for statements that use the syntax `LOCK IN SHARE MODE`. This helps alert users that this feature is not implemented by TiDB, but a workaround to restore the previous behavior is available by setting `tidb_enable_noop_functions=1`. [#21005](https://github.com/pingcap/tidb/pull/21005)
    - Avoid unnecessary warnings/errors when folding constants in shortcut-able expressions [#21040](https://github.com/pingcap/tidb/pull/21040)
    - Raise an error when preparing the `load data` statement. [#21199](https://github.com/pingcap/tidb/pull/21199)
    - Ignore integer zero-fill size attribute when changing the column types [#20986](https://github.com/pingcap/tidb/pull/20986)
    - Add runtime information for DML in explain analyze [#21066](https://github.com/pingcap/tidb/pull/21066)
    - Disallow multi-updates on the primary key (i.e. updating primary key of the same table with different alias) [#21113](https://github.com/pingcap/tidb/pull/21113)
    - Add metric for connection idle time [#21301](https://github.com/pingcap/tidb/pull/21301)
```

        - Exclude system databases in --filter parameter by default [#194](https://github.com/pingcap/dumpling/pull/194)
        - Support --transactional-consistency parameter and support rebuild mysql connections during retry [#199](https://github.com/pingcap/dumpling/pull/199)

    + TiDB Lightning

        - Filter out all system schemas by default [#459](https://github.com/pingcap/tidb-lightning/pull/459)
        - Support set a default value for auto random primary key for local/importer backend [#457](https://github.com/pingcap/tidb-lightning/pull/457)
        - Use range properties to make range split more precise with local backend [#422](https://github.com/pingcap/tidb-lightning/pull/422)
        - `tikv-importer.region-split-size`, `mydumper.read-block-size`, `mydumper.batch-size` and `mydumper.max-region-size` can now accept human-readable format in the form "2.5 GiB" [#471](https://github.com/pingcap/tidb-lightning/pull/471)

    + TiDB Binlog

        - Exit Drainer process with non-zero code if upstream PD is down or apply DDL/DML to downstream failed [#1012](https://github.com/pingcap/tidb-binlog/pull/1012)

## Bug Fixes

+ TiDB

>     - Fix incorrect results when using a prefix index with OR condition [#21287](https://github.com/pingcap/tidb/pull/21287)
>     - Fix a bug cause panic when retry is enable. [#21285](https://github.com/pingcap/tidb/pull/21285)
>     - Fix a bug that checks partition definition according to column type [#21273](https://github.com/pingcap/tidb/pull/21273)
>     - Fix a bug that hash type partition does not check partition name is unique [#21257](https://github.com/pingcap/tidb/pull/21257)
>     - Fix unexpected error when `INSERT` meets index join in some cases. [#21249](https://github.com/pingcap/tidb/pull/21249)
>     - Fix insert value into hash partition table which not int [#21238](https://github.com/pingcap/tidb/pull/21238)
>     - Fix Bigint unsigned column value in CASE WHEN operator is converted to bigint signed #9869 [#21236](https://github.com/pingcap/tidb/pull/21236)
>     - Fix a bug that index-hash-join and index-merge-join doesn't consider collation [#21219](https://github.com/pingcap/tidb/pull/21219)
>     - Fix issue of the query result of slow_query maybe miss some rows. [#21211](https://github.com/pingcap/tidb/pull/21211)
>     - `DELETE` may not delete data correctly when the database name is not in pure lower representation. [#21206](https://github.com/pingcap/tidb/pull/21206)
>     - Raise an error when preparing the `load data` statement. [#21199](https://github.com/pingcap/tidb/pull/21199)
>     - Fix a bug that partition table does not consider collation in `create table` and `select` sentence. [#21181](https://github.com/pingcap/tidb/pull/21181)
>     - Disallow multi-updates on primary key [#21113](https://github.com/pingcap/tidb/pull/21113)
>     - Fix a bug causes schema change after DML [#21050](https://github.com/pingcap/tidb/pull/21050)
>     - Fix the bug that can not query the coalesced column when use using-join. [#21021](https://github.com/pingcap/tidb/pull/21021)
>     - Fix wrong results for some semi join queries. [#21019](https://github.com/pingcap/tidb/pull/21019)
>     - Fix table lock for update statement. [#21002](https://github.com/pingcap/tidb/pull/21002)
>     - Fix stack overflow when building recursive view. [#21001](https://github.com/pingcap/tidb/pull/21001)
>     - Ddl: ignore integer zerofill size attribute when changing the column types [#20986](https://github.com/pingcap/tidb/pull/20986)
>     - Fix unexpected results when do merge join on outer join [#20954](https://github.com/pingcap/tidb/pull/20954)
>     - Fix the issue that sometimes a transaction that has undetermined result may be treated as failed. [#20925](https://github.com/pingcap/tidb/pull/20925)

Reviewer 8: ```suggestion
    - Fix incorrect results when using a prefix index with OR condition [#21287](https://github.com/pingcap/tidb/pull/21287)
    - Fix a bug that causes panic when `retry` is enabled. [#21285](https://github.com/pingcap/tidb/pull/21285)
    - Fix a bug of partition definition checking. The value comparison should be in accord with column type [#21273](https://github.com/pingcap/tidb/pull/21273)
    - Fix a bug of partition table's partition column values type check [#21136](https://github.com/pingcap/tidb/pull/21136)
    - Fix a bug that hash type partition does not check partition name is unique [#21257](https://github.com/pingcap/tidb/pull/21257)
    - Fix insert value into hash partition table which not int [#21238](https://github.com/pingcap/tidb/pull/21238)
    - Fix unexpected error when `INSERT` meets index join in some cases. [#21249](https://github.com/pingcap/tidb/pull/21249)
    - Fix bigint unsigned column value in CASE WHEN operator is converted to bigint signed  [#21236](https://github.com/pingcap/tidb/pull/21236)
    - Fix a bug that index-hash-join and index-merge-join doesn't consider collation [#21219](https://github.com/pingcap/tidb/pull/21219)
    - Fix a bug that partition table does not consider collation in `create table` and `select` sentence. [#21181](https://github.com/pingcap/tidb/pull/21181)
    - Fix issue of the query result of slow_query maybe miss some rows. [#21211](https://github.com/pingcap/tidb/pull/21211)
    - `DELETE` may not delete data correctly when the database name is not in pure lower representation. [#21206](https://github.com/pingcap/tidb/pull/21206)
    - Fix a bug causes schema change after DML [#21050](https://github.com/pingcap/tidb/pull/21050)
    - Fix the bug that can not query the coalesced column when use using-join. [#21021](https://github.com/pingcap/tidb/pull/21021)
    - Fix wrong results for some semi-join queries. [#21019](https://github.com/pingcap/tidb/pull/21019)
    - Fix table lock for update statement. [#21002](https://github.com/pingcap/tidb/pull/21002)
    - Fix stack overflow when building recursive view. [#21001](https://github.com/pingcap/tidb/pull/21001)
    - Fix unexpected results when do merge join on outer join [#20954](https://github.com/pingcap/tidb/pull/20954)
    - Fix the issue that sometimes a transaction that has an undetermined result may be treated as failed. [#20925](https://github.com/pingcap/tidb/pull/20925)
    - Fix issue `explain for connection` can't show the last query plan/ [#21315](https://github.com/pingcap/tidb/pull/21315)
    - Fix the issue that when Index Merge is used in a transaction with RC isolation level, the result may be incorrect. [#21253](https://github.com/pingcap/tidb/pull/21253)
    - Fix auto-id allocation failed because of the transaction's write-conflict retry. [#21079](https://github.com/pingcap/tidb/pull/21079)
    - Fix The JSON Data can not import to TiDB correctly by `load data` [#21074](https://github.com/pingcap/tidb/pull/21074)
    - Set correct default value for new added enum column. [#20998](https://github.com/pingcap/tidb/pull/20998)
    - Expression: keep the original data type when doing date arithmetic operations [#21176](https://github.com/pingcap/tidb/pull/21176)
    - Fix the wrong point get plan generation in fast plan code path [#21244](https://github.com/pingcap/tidb/pull/21244)
```

+ Tools

    + Backup & Restore (BR)

        - Fix special characters in S3 secret access keys [#617](https://github.com/pingcap/br/pull/617)

    + TiCDC

        - Fix the bug that multiple owners could exist when owner campaign key is deleted [#1104](https://github.com/pingcap/ticdc/pull/1104)
        - Kafka_consumer: fix wrong timezone when 'TZ' env is not set and downstream is mysql [#1095](https://github.com/pingcap/ticdc/pull/1095)

    + Dumpling

        - Fix the problem that dumpling may get blocked when its connection to database server is closed [#190](https://github.com/pingcap/dumpling/pull/190)

    + TiDB Lightning

        - Fix a bug about encoding data with wrong field info [#437](https://github.com/pingcap/tidb-lightning/pull/437)
        - Fix a bug that gc life time ttl does not take effect [#448](https://github.com/pingcap/tidb-lightning/pull/448)
        - Fix a bug that causes panic when manually stops Lightning import in Local-backend mode [#484](https://github.com/pingcap/tidb-lightning/pull/484)

## Others

+ TiDB

>     - Fix issue of explain for connection can't show the last query plan/ [#21315](https://github.com/pingcap/tidb/pull/21315)
>     - Add metric for connection idle time [#21301](https://github.com/pingcap/tidb/pull/21301)
>     - Fix the issue that when Index Merge is used in a transaction with RC isolation level, the result may be incorrect. [#21253](https://github.com/pingcap/tidb/pull/21253)
>     - Planner: fix wrong point get plan generation in fast plan code path [#21244](https://github.com/pingcap/tidb/pull/21244)
>     - Expression: keep the original data type when doing date arithmetic operations [#21176](https://github.com/pingcap/tidb/pull/21176)
>     - Ddl: Fix partition table's partition column values type check [#21136](https://github.com/pingcap/tidb/pull/21136)
>     - Make dump binary time more compatible with MySQL [#21135](https://github.com/pingcap/tidb/pull/21135)
>     - Fix auto-id allocation failed because of transaction's write-conflict retry. [#21079](https://github.com/pingcap/tidb/pull/21079)
>     - Fix The JSON Data can not import to TiDB correctly by `load data` [#21074](https://github.com/pingcap/tidb/pull/21074)
>     - Fix up displaying executor runtime info & Add runtime information for DML in explain analyze [#21066](https://github.com/pingcap/tidb/pull/21066)
>     - Deprecate `enable-streaming`. [#21055](https://github.com/pingcap/tidb/pull/21055)
>     - Avoid unnecessary warnings/errors when folding constants in shortcut-able expressions [#21040](https://github.com/pingcap/tidb/pull/21040)
>     - TiDB now provides an error for statements that use the syntax `LOCK IN SHARE MODE`. This helps alert users that this feature is not implemented by TiDB, but a workaround to restore the previous behavior is available by setting `tidb_enable_noop_functions=1`. [#21005](https://github.com/pingcap/tidb/pull/21005)
>     - Set correct default value for new added enum column. [#20998](https://github.com/pingcap/tidb/pull/20998)

Reviewer 8: ```suggestion
    - Deprecate `enable-streaming`. [#21055](https://github.com/pingcap/tidb/pull/21055)
```

        - Support the `--transactional-consistency` parameter and support rebuilding MySQL connections during retry [#199](https://github.com/pingcap/dumpling/pull/199)

    + TiDB Lightning

        - Filter out all system schemas by default [#459](https://github.com/pingcap/tidb-lightning/pull/459)
        - Support setting a default value for the auto-random primary key for the Local-backend or Importer-backend [#457](https://github.com/pingcap/tidb-lightning/pull/457)
        - Use range properties to make the range split more precise in Local-backend [#422](https://github.com/pingcap/tidb-lightning/pull/422)
        - Support a human-readable format (such as "2.5 GiB") in `tikv-importer.region-split-size`, `mydumper.read-block-size`, `mydumper.batch-size`, and `mydumper.max-region-size` [#471](https://github.com/pingcap/tidb-lightning/pull/471)

    + TiDB Binlog

        - Exit the Drainer process with the non-zero code if the upstream PD is down or if applying DDL or DML statements to the downstream fails [#1012](https://github.com/pingcap/tidb-binlog/pull/1012)

## Bug Fixes

> + TiDB

Reviewer 9: ```
- Fix a bug that delete operation does not lock unique key [#20705](https://github.com/pingcap/tidb/pull/20705)
- Fix a bug that snapshot read will hit lock cache [#21539](https://github.com/pingcap/tidb/pull/21539)
```

Reviewer 9: the corresponding change in docs-cn is https://github.com/pingcap/docs-cn/pull/4976#discussion_r542411298

> 

Reviewer 10: - Fix the issue of `ADD_DATE` adjusting to daylight saving time change [#20888](https://github.com/pingcap/tidb/pull/20888)
- Fix a bug that prevents insertion of string with trainling spaces that exceed `varchar` or `char`'s length constraint [#21282](https://github.com/pingcap/tidb/pull/21282)
- Fix a bug about comparing `int` with `year` that does not converting int from [1, 69] to [2001, 2069] and from [70, 99] to [1970, 1999] [#21283](https://github.com/pingcap/tidb/pull/21283)

Reviewer 10: the corresponding change in docs-cn is https://github.com/pingcap/docs-cn/pull/4976#discussion_r542371378

    - Fix the unexpected result returned when performing index merge join operations on outer join [#20954](https://github.com/pingcap/tidb/pull/20954)
    - Fix the issue that sometimes a transaction that has an undetermined result might be treated as failed [#20925](https://github.com/pingcap/tidb/pull/20925)
    - Fix the issue that `EXPLAIN FOR CONNECTION` cannot show the last query plan [#21315](https://github.com/pingcap/tidb/pull/21315)
    - Fix the issue that when Index Merge is used in a transaction with the Read Committed isolation level, the result might be incorrect [#21253](https://github.com/pingcap/tidb/pull/21253)
    - Fix the auto-ID allocation failure caused by the transaction retry after the write conflict [#21079](https://github.com/pingcap/tidb/pull/21079)
    - Fix the issue that JSON data cannot be correctly imported to TiDB using `LOAD DATA` [#21074](https://github.com/pingcap/tidb/pull/21074)
    - Fix the issue that the default value of newly added `Enum`-type columns is incorrect [#20998](https://github.com/pingcap/tidb/pull/20998)

>     - Fix the issue that the `addata` function inserts invalid characters [#21176](https://github.com/pingcap/tidb/pull/21176)

Reviewer 8: ```suggestion
    - Fix the issue that the `adddate` function inserts invalid characters [#21176](https://github.com/pingcap/tidb/pull/21176)
```

    - Fix the issue that the wrong `PointGet` plan generated in some situations causes wrong results [#21244](https://github.com/pingcap/tidb/pull/21244)

> 

Reviewer 11: -Fixed the problem of panic caused by the overflow of the result of the sum() function calculating the double type field [#21272](https://github.com/pingcap/tidb/pull/21272)

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

> Release date: December 15, 2020

Reviewer 2: ```suggestion
Release date: December 18, 2020
```

    - Fix the issue that the wrong `PointGet` plan generated in some situations causes wrong results [#21244](https://github.com/pingcap/tidb/pull/21244)
    - Ignore the conversion of daylight saving time in the `ADD_DATE` function to be compatible with MySQL [#20888](https://github.com/pingcap/tidb/pull/20888)
    - Fix a bug that prevents inserting strings with trailing spaces that exceed `varchar` or `char`'s length constraint [#21282](https://github.com/pingcap/tidb/pull/21282)
    - Fix a bug that does not converting the integer from `[1, 69]` to `[2001, 2069]` or from `[70, 99]` to `[1970, 1999]` when comparing `int` with `year` [#21283](https://github.com/pingcap/tidb/pull/21283)
    - Fix the panic caused by the overflowing result of the `sum()` function when calculating the `Double` type field [#21272](https://github.com/pingcap/tidb/pull/21272)
    - Fix a bug that `DELETE` fails to add lock on the unique key [#20705](https://github.com/pingcap/tidb/pull/20705)
    - Fix a bug that snapshot reads hits the lock cache [#21539](https://github.com/pingcap/tidb/pull/21539)

> 

Reviewer 12: ```suggestion
    - Fix a potential memory leak after read lots of data in a long-lived transaction [#21129](https://github.com/pingcap/tidb/pull/21129)

```

> 

Reviewer 13: ```suggestion
    - Fix the issue that omitting table alias in subquery reports syntax error [#20367](https://github.com/pingcap/tidb/pull/20367)
```

Reviewer 2: This note is not added in https://github.com/pingcap/docs-cn/pull/4976.

Reviewer 13: Updated: https://github.com/pingcap/docs-cn/pull/4976/files#r545553343

        - Fix a bug that TiCDC might fail to continue replicating data when a TiKV node crashes or recovers from a crash. This bug only exists in v4.0.8. [#1198](https://github.com/pingcap/ticdc/pull/1198)
        - Fix the issue that the metadata is repeatedly flushed to etcd before a table is initialized [#1191](https://github.com/pingcap/ticdc/pull/1191)
        - Fix an issue of replication interruption caused by early GC or the latency of updating `TableInfo` when the schema storage caches TiDB tables [#1114](https://github.com/pingcap/ticdc/pull/1114)
        - Fix the issue that the schema storage costs too much memory when DDL operations are frequent [#1127](https://github.com/pingcap/ticdc/pull/1127)
        - Fix the goroutine leak when a changefeed is paused or stopped [#1075](https://github.com/pingcap/ticdc/pull/1075)
        - Increase the maximum retry timeout to 60 seconds in Kafka producer to prevent replication interruption caused by the service or network jitter in the downstream Kafka [#1118](https://github.com/pingcap/ticdc/pull/1118)
        - Fix a bug that the Kafka batch size does not take effect [#1112](https://github.com/pingcap/ticdc/pull/1112)

> 

Reviewer 7: ```suggestion
        - Fix a bug that some tables' row change could be lost when the network between CDC and PD has jitter, and there exists changefeed resume operation at the same time [#1213](https://github.com/pingcap/ticdc/pull/1213)

```

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

Release date: December 18, 2020

TiDB version: 4.0.9

## Compatibility Changes

+ TiDB

    - Deprecate the `enable-streaming` configuration item [#21055](https://github.com/pingcap/tidb/pull/21055)

> 

Reviewer: ```suggestion
+ TiKV
        - Reduce IO and mutex contention when encryption-at-rest is enabled. The change is backward incompatible. If user need to downgrade to versions prior to 4.0.9, security.encryption.enable-file-dictionary-log has to be turned off and TiKV has to be restarted once before being downgrade. [#9195](https://github.com/tikv/tikv/pull/9195)

```

## New Features

+ TiFlash

    - Support storing the latest data of the storage engine on multiple disks (experimental)

> 

Reviewer 14: ```suggestion

+ TiDB Dashboard

  - Support displaying and sorting by all fields in SQL Statement [#749](https://github.com/pingcap/tidb-dashboard/pull/749)
  - Support zooming and paning topology graph [#772](https://github.com/pingcap/tidb-dashboard/pull/772)
  - Support displaying disk usage information in SQL Statement and Slow Query [#777](https://github.com/pingcap/tidb-dashboard/pull/777)
  - Support exporting list data for SQL Statement and Slow Query [#778](https://github.com/pingcap/tidb-dashboard/pull/778)
  - Support customizing Prometheus address [#808](https://github.com/pingcap/tidb-dashboard/pull/808)
  - Add cluster statistics page [#815](https://github.com/pingcap/tidb-dashboard/pull/815)
  - Add more time related fields in Slow Query details [#810](https://github.com/pingcap/tidb-dashboard/pull/810)
 
```

+ TiKV

    - Add the tag to trace the source of the `split` command [#8936](https://github.com/tikv/tikv/pull/8936)
    - Support dynamically changing the `pessimistic-txn.pipelined` configuration [#9100](https://github.com/tikv/tikv/pull/9100)
    - Reduce the impact on performance when running Backup & Restore and TiDB Lightning [#9098](https://github.com/tikv/tikv/pull/9098)
    - Add monitoring metrics for the ingesting SST errors [#9096](https://github.com/tikv/tikv/pull/9096)
    - Prevent the leader from being hibernated when some peers still need to replicate logs [#9093](https://github.com/tikv/tikv/pull/9093)
    - Increase the success rate of the pipelined pessimistic locking [#9086](https://github.com/tikv/tikv/pull/9086)
    - Change the default value of `apply-max-batch-size` and `store-max-batch-size` to `1024` [#9020](https://github.com/tikv/tikv/pull/9020)
    - Add the `max-background-flushes` configuration item [#8947](https://github.com/tikv/tikv/pull/8947)
    - Disable `force-consistency-checks` by default to improve performance [#9029](https://github.com/tikv/tikv/pull/9029)

> 

Reviewer: ```suggestion
    - Reduce IO and mutex contention when encryption-at-rest is enabled. The change is backward incompatible. If user need to downgrade to versions prior to 4.0.9, security.encryption.enable-file-dictionary-log has to be turned off and TiKV has to be restarted once before being downgrade. [#9195](https://github.com/tikv/tikv/pull/9195)
    - Offload region size query from pd heartbeat worker to split check worker [#9185](https://github.com/tikv/tikv/pull/9185)

```

+ PD

    - Check the TiKV cluster version when a TiKV stores become `Tombstone`, which prevents users from enabling incompatible features during the process of downgrade or upgrade [#3213](https://github.com/pingcap/pd/pull/3213)
    - Disallow the TiKV store of a lower version to change from `Tombstone` back to `Up` [#3206](https://github.com/pingcap/pd/pull/3206)

>     - Update Dashboard to `v2020.11.26.1` [#3219](https://github.com/pingcap/pd/pull/3219)

Reviewer 14: ```suggestion
```

+ TiFlash

    - Reduce the latency of replica reads
    - Refine TiFlash's error messages
    - Limit the memory usage of cache data when the data volume is huge
    - Add a monitoring metric for the number of coprocessor tasks being handled

> + Tools

Reviewer 14: ```suggestion
+ TiDB Dashboard

  - Keep expanding when "Expand" is clicked for SQL statements [#775](https://github.com/pingcap/tidb-dashboard/pull/775)
  - Open detail pages in new window for SQL Statement and Slow Query [#816](https://github.com/pingcap/tidb-dashboard/pull/816)
  - Improve descriptions for time related fields in Slow Query details [#817](https://github.com/pingcap/tidb-dashboard/pull/817)
  - Display detailed error messages [#794](https://github.com/pingcap/tidb-dashboard/pull/794)
  
+ Tools
```

    - Fix the issue that Region Merge might cause data loss during network partition [#9108](https://github.com/tikv/tikv/pull/9108)
    - Fix the issue that the `ANALYZE` statement might cause panic when using the `latin1` character set [#9082](https://github.com/tikv/tikv/pull/9082)
    - Fix the wrong results returned when converting the numeric type to the time type [#9031](https://github.com/tikv/tikv/pull/9031)
    - Fix a bug that TiDB Lightning fails to ingest SST files to TiKV with the Importer-backend or Local-backend when Transparent Data Encryption (TDE) is enabled [#8995](https://github.com/tikv/tikv/pull/8995)
    - Fix the invalid `advertise-status-addr` value (`0.0.0.0`) [#9036](https://github.com/tikv/tikv/pull/9036)
    - Fix the issue that an error is returned indicating that a key exists when this key is locked and deleted in a committed transaction [#8930](https://github.com/tikv/tikv/pull/8930)
    - Fix the issue that the RocksDB cache mapping error causes data corruption [#9029](https://github.com/tikv/tikv/pull/9029)
    - Fix a bug that Follower Read might return stale data after the leader is transferred [#9240](https://github.com/tikv/tikv/pull/9240)

> 

Reviewer: ```suggestion
    - Fix stale old value with pessimistic lock [#9282](https://github.com/tikv/tikv/pull/9282)
    - Fix the bug that follower read may get stale data after transferring leader [#9240](https://github.com/tikv/tikv/pull/9240)
    - Fix TiKV crash when receiving SIGPROF after profiling [#9229](https://github.com/tikv/tikv/pull/9229)

```

+ PD

    - Fix the issue that the leader roles specified using placement rules do not take effect in some cases [#3208](https://github.com/pingcap/pd/pull/3208)
    - Fix the issue that the `trace-region-flow` value is unexpectedly set to `false` [#3120](https://github.com/pingcap/pd/pull/3120)
    - Fix a bug that the service safepoint with infinite Time To Live (TTL) does not work [#3143](https://github.com/pingcap/pd/pull/3143)

+ TiFlash

    - Fix the issue that `INFORMATION_SCHEMA.CLUSTER_HARDWARE` might contain the information of disks that are not in use
    - Fix the issue that the estimate on memory usage of Delta Cache is smaller than the actual usage
    - Fix the memory leak caused by thread information statistics

> + Tools

Reviewer 14: ```suggestion
+ TiDB Dashboard

  - Fix relative time translation in Chinese language [#755](https://github.com/pingcap/tidb-dashboard/pull/755)
  - Fix a bug that browser compatibility notice is not working [#776](https://github.com/pingcap/tidb-dashboard/pull/776)
  - Fix incorrect transaction start_ts in some scenarios [#793](https://github.com/pingcap/tidb-dashboard/pull/793)
  - Fix incorrect formatting some SQL text [#805](https://github.com/pingcap/tidb-dashboard/pull/805)
  
+ Tools
```

    + Backup & Restore (BR)

        - Fix the failure caused by special characters in S3 secret access keys [#617](https://github.com/pingcap/br/pull/617)

    + TiCDC

        - Fix the issue that multiple owners might exist when the owner campaign key is deleted [#1104](https://github.com/pingcap/ticdc/pull/1104)
        - Fix a bug that TiCDC might fail to continue replicating data when a TiKV node crashes or recovers from a crash. This bug only exists in v4.0.8. [#1198](https://github.com/pingcap/ticdc/pull/1198)
        - Fix the issue that the metadata is repeatedly flushed to etcd before a table is initialized [#1191](https://github.com/pingcap/ticdc/pull/1191)
        - Fix an issue of replication interruption caused by early GC or the latency of updating `TableInfo` when the schema storage caches TiDB tables [#1114](https://github.com/pingcap/ticdc/pull/1114)
        - Fix the issue that the schema storage costs too much memory when DDL operations are frequent [#1127](https://github.com/pingcap/ticdc/pull/1127)
        - Fix the goroutine leak when a changefeed is paused or stopped [#1075](https://github.com/pingcap/ticdc/pull/1075)

>         - Increase the maximum retry timeout to 60 seconds in Kafka producer to prevent replication interruption caused by the service or network jitter in the downstream Kafka [#1118](https://github.com/pingcap/ticdc/pull/1118)

Reviewer 2: ```suggestion
        - Increase the maximum retry timeout to 600 seconds in Kafka producer to prevent replication interruption caused by the service or network jitter in the downstream Kafka [#1118](https://github.com/pingcap/ticdc/pull/1118)
```

        - Fix a bug that the Kafka batch size does not take effect [#1112](https://github.com/pingcap/ticdc/pull/1112)
        - Fix a bug that some tables' row change might be lost when the network between TiCDC and PD has jitter and when there are paused changefeeds being resumed at the same time [#1213](https://github.com/pingcap/ticdc/pull/1213)
        - Fix a bug that the TiCDC process might exit when the network between TiCDC and PD is not stable [#1218](https://github.com/pingcap/ticdc/pull/1218)
        - Use a singleton PD client in TiCDC and fix a bug that TiCDC closes PD client by accident which causes replication block [#1217](https://github.com/pingcap/ticdc/pull/1217)

>         - Fix a bug that the cdc owner might consume too much memory in the etcd watch client

Reviewer 7: ```suggestion
        - Fix a bug that the TiCDC owner might consume too much memory in the etcd watch client [#1224](https://github.com/pingcap/ticdc/pull/1224)
```

---
title: TiDB 4.0.9 Release Notes
---

# TiDB 4.0.9 Release Notes

> Release date: December 18, 2020

Reviewer 2: ```suggestion
Release date: December 21, 2020
```
